File: app/aria2.py
# -*- coding: utf-8 -*-
import requests
import time
from pathlib import Path
from typing import Dict, Optional

from .config import state
from .utils import log, ensure_dir, bytes_to_speed

class Aria2RpcClient:
    def __init__(self, rpc_url: str, secret: str):
        self.url = rpc_url
        self.secret = secret
        self.session = requests.Session()

    def _rpc_call(self, method: str, params: list) -> Dict:
        payload = {
            "jsonrpc": "2.0",
            "id": "emos",
            "method": method,
            "params": [f"token:{self.secret}"] + params,
        }
        try:
            r = self.session.post(self.url, json=payload, timeout=10)
            r.raise_for_status()
            return r.json()
        except requests.RequestException as e:
            log(f"Aria2 RPC call failed: {e}", "ERROR")
            return {"error": str(e)}

    def add_download(self, url: str, options: Dict) -> Optional[str]:
        res = self._rpc_call("aria2.addUri", [[url], options])
        return res.get("result")

    def get_status(self, gid: str) -> Optional[Dict]:
        res = self._rpc_call("aria2.tellStatus", [gid])
        return res.get("result")

    def remove_download(self, gid: str, force: bool = False):
        if force:
            self._rpc_call("aria2.forceRemove", [gid])
        else:
            self._rpc_call("aria2.removeDownloadResult", [gid])

    def check_version(self) -> bool:
        res = self._rpc_call("aria2.getVersion", [])
        if "result" in res:
            log(f"Aria2 connected: version {res['result']['version']}", "INFO")
            return True
        log(f"Aria2 connection failed: {res.get('error')}", "ERROR")
        return False

    def download_and_monitor(self, url: str, dst_path: str, threads: int) -> bool:
        ensure_dir(str(Path(dst_path).parent))
        options = {
            "dir": str(Path(dst_path).parent),
            "out": Path(dst_path).name,
            "allow-overwrite": "true",
            "auto-file-renaming": "false",
            "split": str(threads),
            "max-connection-per-server": str(threads),
        }
        gid = self.add_download(url, options)
        if not gid:
            log("Failed to add download task to aria2", "ERROR")
            return False

        log(f"Aria2 task added: GID={gid}", "INFO")

        while True:
            if state.task["cancel"]:
                log("Cancellation received, removing aria2 task", "WARN")
                self.remove_download(gid, force=True)
                return False

            status = self.get_status(gid)
            if not status:
                time.sleep(2)
                continue

            total = int(status.get("totalLength", 0))
            completed = int(status.get("completedLength", 0))
            speed = int(status.get("downloadSpeed", 0))
            percent = (completed / total * 100) if total > 0 else 0
            eta_seconds = (total - completed) / speed if speed > 0 else 0

            state.task["download"].update({
                "percent": percent,
                "speed": bytes_to_speed(speed),
                "eta": f"{int(eta_seconds // 60)}m {int(eta_seconds % 60)}s" if eta_seconds > 0 else "N/A",
            })

            if status["status"] == "complete":
                state.task["download"].update({"percent": 100.0, "speed": "done", "done": True})
                log("Aria2 download complete", "SUCCESS")
                self.remove_download(gid)
                return True
            elif status["status"] in ("error", "removed"):
                error_msg = status.get("errorMessage", "Unknown error")
                log(f"Aria2 download failed: {error_msg}", "ERROR")
                self.remove_download(gid, force=True)
                return False

            time.sleep(1)

File: goserver/internal/service/aria2.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"path/filepath"
	"time"
)

type Aria2Service struct {
	client *http.Client
}

type Aria2Progress struct {
	Percent float64
	Speed   string
	ETA     string
	Done    bool
}

func NewAria2Service() *Aria2Service {
	return &Aria2Service{client: &http.Client{Timeout: 10 * time.Second}}
}

func (s *Aria2Service) CheckVersion(rpcURL, secret string) error {
	res, err := s.rpcCall(rpcURL, secret, "aria2.getVersion", []any{})
	if err != nil {
		return err
	}
	if _, ok := res["result"]; !ok {
		return fmt.Errorf("aria2 connection failed")
	}
	return nil
}

func (s *Aria2Service) DownloadAndMonitor(ctx context.Context, rpcURL, secret, url, dstPath string, threads int, onProgress func(Aria2Progress)) error {
	options := map[string]string{
		"dir":                       filepath.Dir(dstPath),
		"out":                       filepath.Base(dstPath),
		"allow-overwrite":           "true",
		"auto-file-renaming":        "false",
		"split":                     fmt.Sprintf("%d", threads),
		"max-connection-per-server": fmt.Sprintf("%d", threads),
	}
	res, err := s.rpcCall(rpcURL, secret, "aria2.addUri", []any{[]string{url}, options})
	if err != nil {
		return err
	}
	gid, _ := res["result"].(string)
	if gid == "" {
		return fmt.Errorf("failed to add download task to aria2")
	}

	for {
		select {
		case <-ctx.Done():
			_, _ = s.rpcCall(rpcURL, secret, "aria2.forceRemove", []any{gid})
			return ctx.Err()
		case <-time.After(1 * time.Second):
		}

		status, err := s.rpcCall(rpcURL, secret, "aria2.tellStatus", []any{gid})
		if err != nil {
			continue
		}
		result, _ := status["result"].(map[string]any)
		if result == nil {
			continue
		}

		total := anyToInt64(result["totalLength"])
		completed := anyToInt64(result["completedLength"])
		speed := anyToInt64(result["downloadSpeed"])
		percent := 0.0
		if total > 0 {
			percent = float64(completed) / float64(total) * 100
		}
		eta := "N/A"
		if speed > 0 {
			remain := (total - completed) / speed
			eta = fmt.Sprintf("%dm %ds", remain/60, remain%60)
		}
		if onProgress != nil {
			onProgress(Aria2Progress{
				Percent: percent,
				Speed:   bytesToSpeed(speed),
				ETA:     eta,
				Done:    false,
			})
		}

		state, _ := result["status"].(string)
		switch state {
		case "complete":
			_, _ = s.rpcCall(rpcURL, secret, "aria2.removeDownloadResult", []any{gid})
			if onProgress != nil {
				onProgress(Aria2Progress{Percent: 100, Speed: "done", ETA: "N/A", Done: true})
			}
			return nil
		case "error", "removed":
			_, _ = s.rpcCall(rpcURL, secret, "aria2.forceRemove", []any{gid})
			return fmt.Errorf("aria2 download failed")
		}
	}
}

func (s *Aria2Service) rpcCall(rpcURL, secret, method string, params []any) (map[string]any, error) {
	rpcParams := append([]any{}, params...)
	if secret != "" {
		rpcParams = append([]any{"token:" + secret}, rpcParams...)
	}
	payload := map[string]any{
		"jsonrpc": "2.0",
		"id":      "goserver",
		"method":  method,
		"params":  rpcParams,
	}
	body, _ := json.Marshal(payload)
	resp, err := s.client.Post(rpcURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var out map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

func anyToInt64(v any) int64 {
	switch x := v.(type) {
	case string:
		var out int64
		fmt.Sscan(x, &out)
		return out
	case float64:
		return int64(x)
	case int64:
		return x
	case int:
		return int64(x)
	default:
		return 0
	}
}

func bytesToSpeed(v int64) string {
	mb := float64(v) / 1024.0 / 1024.0
	if mb >= 1 {
		return fmt.Sprintf("%.2f MB/s", mb)
	}
	kb := float64(v) / 1024.0
	return fmt.Sprintf("%.2f KB/s", kb)
}
